core/src/Cabana_ParticleList.hpp

```diff
@@ -0,0 +1,249 @@
+{
+    return Cabana::get<TypeIndexer<FieldTag, FieldTags...>::index>(
+        particle.tuple(), indices... );
+}
+
+//---------------------------------------------------------------------------//
+//! Return a single SoA for a single particle.
+template <class FieldTag, class... FieldTags, class... IndexTypes,
+          int VectorLength>
+KOKKOS_FORCEINLINE_FUNCTION typename std::enable_if<
+    sizeof...( IndexTypes ) == FieldTag::rank,
+    typename ParticleView<VectorLength, FieldTags...>::soa_type::
+        template member_const_reference_type<
+            TypeIndexer<FieldTag, FieldTags...>::index>>::type
+get( const ParticleView<VectorLength, FieldTags...>& particle, FieldTag,
+     IndexTypes... indices )
+{
+    return Cabana::get<TypeIndexer<FieldTag, FieldTags...>::index>(
+        particle.soa(), particle.vectorIndex(), indices... );
+}
+
+//! Return a single SoA for a single particle.
+template <class FieldTag, class... FieldTags, class... IndexTypes,
+          int VectorLength>
+KOKKOS_FORCEINLINE_FUNCTION typename std::enable_if<
+    sizeof...( IndexTypes ) == FieldTag::rank,
+    typename ParticleView<VectorLength, FieldTags...>::soa_type::
+        template member_reference_type<
+            TypeIndexer<FieldTag, FieldTags...>::index>>::type
+get( ParticleView<VectorLength, FieldTags...>& particle, FieldTag,
+     IndexTypes... indices )
+{
+    return Cabana::get<TypeIndexer<FieldTag, FieldTags...>::index>(
+        particle.soa(), particle.vectorIndex(), indices... );
+}
+
+//---------------------------------------------------------------------------//
+//! List of particle fields stored in AoSoA.
+template <class MemorySpace, class... FieldTags>
+class ParticleList
+{
+  public:
+    //! Kokkos memory space.
+    using memory_space = MemorySpace;
+    //! Particle AoSoA member types.
+    using traits = ParticleTraits<FieldTags...>;
+    //! AoSoA type.
+    using aosoa_type =
+        Cabana::AoSoA<typename traits::member_types, memory_space>;
+    //! Particle tuple type.
+    using tuple_type = typename aosoa_type::tuple_type;
+    /*!
+      \brief Single field slice type.
+      \tparam M AoSoA field index.
+    */
+    template <std::size_t M>
+    using slice_type = typename aosoa_type::template member_slice_type<M>;
+    //! Single particle type.
+    using particle_type = Particle<FieldTags...>;
+    //! Single SoA type.
+    using particle_view_type =
+        ParticleView<aosoa_type::vector_length, FieldTags...>;
+
+    //! Default constructor.
+    ParticleList( const std::string& label )
+        : _aosoa( label )
+        , _label( label )
+    {
+    }
+
+    //! Get the number of particles in the list.
+    std::size_t size() const { return _aosoa.size(); }
+
+    //! Get the AoSoA.
+    aosoa_type& aosoa() { return _aosoa; }
+    //! Get the AoSoA (const).
+    const aosoa_type& aosoa() const { return _aosoa; }
+
+    //! Get the label.
+    const std::string& label() { return _label; }
```

Comment:
should only need to `const` version of this function because you always return a `const &`

Reply:
Fixed